### ISSUE TITLE
Enabling CRL checks in combination with `cacerts` option causes function_clause error

### ISSUE THREAD

**Issue by voltone** [erlang/otp#4589]
**Describe the bug**
I have seen TLS handshakes fail with `{internal_error, "TLS client: In state wait_cert_cr at tls_handshake_1_3.erl:1428 generated CLIENT ALERT: Fatal - Internal Error\n {unexpected_error,function_clause}"}` when enabling CRL validation. I have traced down the error to a call to `ssl_crl:trusted_cert_and_path/3` with the following arguments:

1. The CRL record
2. `issuer_not_found`
3. `{#Ref<0.2415818995.3577348099.85245>, {extracted, [{decoded,...}]}}`

The 3rd argument looks like a `DbInfo` tuple, which is supposed to be handled by the 3rd clause of `ssl_crl:trusted_cert_and_path/3` on line 49, but it also matches the pattern `{CertPath, {Db, DbRef}}` on line 40. As a result, the second function clause is invoked, which subsequently failed with said error, since `CertPath` is a reference rather than a list.

**To Reproduce**
The issue is not easily reproducible as it requires a set of certificates and CRLs. I will try to produce those and attach them to the report at a later time.

**Expected behavior**
No `function_clause` error should be raised during CRL verification. E.g. ensure the `CertPath` element in the 3rd argument in `lib/ssl/src/ssl_crl.erl:40` is a list.

**Affected versions**
I have seen the issue with 23.2.7 and 23.1.2, but not with 23.0.4 or earlier.

**Comment by voltone** [comment]
To reproduce (output shown for OTP 23.2.7):

```erl
Erlang/OTP 23 [erts-11.1.8] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]

Eshell V11.1.8  (abort with ^G)
1> PEM = <<"-----BEGIN CERTIFICATE-----
1> MIIDTjCCAjagAwIBAgIIOp7H0EUnT5YwDQYJKoZIhvcNAQELBQAwMzEfMB0GA1UE
1> CgwWRWxpeGlyLlg1MDkuVGVzdC5TdWl0ZTEQMA4GA1UEAwwHUm9vdCBDQTAeFw0y
1> MTAzMDcxNDAyMTdaFw00NjAzMDcxNDA3MTdaMDMxHzAdBgNVBAoMFkVsaXhpci5Y
1> NTA5LlRlc3QuU3VpdGUxEDAOBgNVBAMMB1Jvb3QgQ0EwggEiMA0GCSqGSIb3DQEB
1> AQUAA4IBDwAwggEKAoIBAQC8JQiK7tIVTGC0qg0NG95wbEsiU2iVTtSwvGSpY/U7
1> wJVyWtxGu2kop6oZ8JVyb1pfWnCHB1p5EV0cvkHtYtxMlhZkot99vgBZNey6VnkT
1> 0SbPVkuTrQnJPFl6cb3DvB+0afs4lOuOV7u17kvgT2Jo8zp28BxrCvov5ZRK2W25
1> Kz6QGkaJHrtprD9iF7iH9byJr8iq/PKDDtuP/U0EgSTp0QNUwxMD8a2Ux7C+CrRm
1> Zo2RPxJ28AELN1kaifi6DxcvAonk8xGhgwoA2xq+ZbAmKfX6qcwUH6OJ7QNoOLOH
1> 6vaddtsCn9JJq3ovm9LH7BIln4Xa8TVzYNU9N6OhHqvTAgMBAAGjZjBkMBIGA1Ud
1> EwEB/wQIMAYBAf8CAQEwDgYDVR0PAQH/BAQDAgGGMB0GA1UdDgQWBBS/ZjHRK+UD
1> Ri5qcbRjJpClA/vOXjAfBgNVHSMEGDAWgBS/ZjHRK+UDRi5qcbRjJpClA/vOXjAN
1> BgkqhkiG9w0BAQsFAAOCAQEAFYJ2Tmb6VYQnt0d9D/Ztd20gty4/HlcGxc1jeiiL
1> ugHdK6DAbdxttlVkT6qOrqWm4EFUgt+ajHTxRZq17Twxa84lBlC+tTCTgjsAli9f
1> /QsLqPCfvVWCCfiZP7C4o61imgktEaNHWIhLXldQBlGKyXGft+PY54MxEmNCyRR/
1> MDEODnjjgp//yIzS/wA1DiE4t4lp9rxbOiaZgjFz7w/ibU6G2a/jJP0i37j2V1d/
1> YFGw4VlDADXyCYrKqaESjCii8AFRPTOJ1qXgjyr85zrN/Kwnus+HMBpHPgiYFH38
1> zPexDUP9sZjhwuY9hNBvcM/9MvIgqv9OxT5BKOY91X9hNA==
1> -----END CERTIFICATE-----">>. 
<<"-----BEGIN CERTIFICATE-----\nMIIDTjCCAjagAwIBAgIIOp7H0EUnT5YwDQYJKoZIhvcNAQELBQAwMzEfMB0GA1UE\nCgwWRWxpeGlyLlg1MDkuVGV"...>>
2> [{'Certificate', DER, _}] = public_key:pem_decode(PEM).
[{'Certificate',<<48,130,3,78,48,130,2,54,160,3,2,1,2,2,8,
                  58,158,199,208,69,39,79,150,48,13,6,...>>,
                not_encrypted}]
3> ssl:start().                                           
ok
4> ssl:connect("revoked.voltone.net", 443, [{verify, verify_peer}, {cacerts, [DER]}, {crl_check, true}, {crl_cache, {ssl_crl_cache, {internal, [{http, 30_000}]}}}]).
=NOTICE REPORT==== 7-Mar-2021::16:38:09.114713 ===
TLS client: In state wait_cert_cr at tls_handshake_1_3.erl:1428 generated CLIENT ALERT: Fatal - Internal Error
 - {unexpected_error,function_clause}
{error,{tls_alert,{internal_error,"TLS client: In state wait_cert_cr at tls_handshake_1_3.erl:1428 generated CLIENT ALERT: Fatal - Internal Error\n {unexpected_error,function_clause}"}}}
5> 
```

For copy & paste convenience, here's the PEM certificate again:

```
-----BEGIN CERTIFICATE-----
MIIDTjCCAjagAwIBAgIIOp7H0EUnT5YwDQYJKoZIhvcNAQELBQAwMzEfMB0GA1UE
CgwWRWxpeGlyLlg1MDkuVGVzdC5TdWl0ZTEQMA4GA1UEAwwHUm9vdCBDQTAeFw0y
MTAzMDcxNDAyMTdaFw00NjAzMDcxNDA3MTdaMDMxHzAdBgNVBAoMFkVsaXhpci5Y
NTA5LlRlc3QuU3VpdGUxEDAOBgNVBAMMB1Jvb3QgQ0EwggEiMA0GCSqGSIb3DQEB
AQUAA4IBDwAwggEKAoIBAQC8JQiK7tIVTGC0qg0NG95wbEsiU2iVTtSwvGSpY/U7
wJVyWtxGu2kop6oZ8JVyb1pfWnCHB1p5EV0cvkHtYtxMlhZkot99vgBZNey6VnkT
0SbPVkuTrQnJPFl6cb3DvB+0afs4lOuOV7u17kvgT2Jo8zp28BxrCvov5ZRK2W25
Kz6QGkaJHrtprD9iF7iH9byJr8iq/PKDDtuP/U0EgSTp0QNUwxMD8a2Ux7C+CrRm
Zo2RPxJ28AELN1kaifi6DxcvAonk8xGhgwoA2xq+ZbAmKfX6qcwUH6OJ7QNoOLOH
6vaddtsCn9JJq3ovm9LH7BIln4Xa8TVzYNU9N6OhHqvTAgMBAAGjZjBkMBIGA1Ud
EwEB/wQIMAYBAf8CAQEwDgYDVR0PAQH/BAQDAgGGMB0GA1UdDgQWBBS/ZjHRK+UD
Ri5qcbRjJpClA/vOXjAfBgNVHSMEGDAWgBS/ZjHRK+UDRi5qcbRjJpClA/vOXjAN
BgkqhkiG9w0BAQsFAAOCAQEAFYJ2Tmb6VYQnt0d9D/Ztd20gty4/HlcGxc1jeiiL
ugHdK6DAbdxttlVkT6qOrqWm4EFUgt+ajHTxRZq17Twxa84lBlC+tTCTgjsAli9f
/QsLqPCfvVWCCfiZP7C4o61imgktEaNHWIhLXldQBlGKyXGft+PY54MxEmNCyRR/
MDEODnjjgp//yIzS/wA1DiE4t4lp9rxbOiaZgjFz7w/ibU6G2a/jJP0i37j2V1d/
YFGw4VlDADXyCYrKqaESjCii8AFRPTOJ1qXgjyr85zrN/Kwnus+HMBpHPgiYFH38
zPexDUP9sZjhwuY9hNBvcM/9MvIgqv9OxT5BKOY91X9hNA==
-----END CERTIFICATE-----
```

Please note that I set up a temporary CRL and HTTPS server for this, but I do not intend to keep that server running indefinitely. So please let me know when you have reproduced the issue so I can take it down. Thanks :)

**Comment by IngelaAndin** [comment]
Thank you for reporting this. The ssl_crl.erl needs an update due to the refactor made for integrating code for future OCSP support. Alas the CRL suite only uses files as input so we missed it. We will update the test suite to include a DER input test too.  

Here is the patch:
```
diff --git a/lib/ssl/src/ssl_crl.erl b/lib/ssl/src/ssl_crl.erl
index 888a75bfd6..0cf1380131 100644
--- a/lib/ssl/src/ssl_crl.erl
+++ b/lib/ssl/src/ssl_crl.erl
@@ -37,7 +37,7 @@ trusted_cert_and_path(CRL, {SerialNumber, Issuer},{_, {Db, DbRef}} = DbHandle) -
            {ok, Root, Chain} = ssl_certificate:certificate_chain(OtpCert, Db, DbRef),
            {ok, Root,  lists:reverse(Chain)}
     end;
-trusted_cert_and_path(CRL, issuer_not_found, {CertPath, {Db, DbRef}}) -> 
+trusted_cert_and_path(CRL, issuer_not_found, {CertPath, {Db, DbRef}}) when is_list(CertPath) -> 
     case find_issuer(CRL, {certpath, 
                            [{Der, public_key:pkix_decode_cert(Der,otp)} || Der <-  CertPath]}) of
        {ok, OtpCert} ->
@@ -46,10 +46,10 @@ trusted_cert_and_path(CRL, issuer_not_found, {CertPath, {Db, DbRef}}) ->
        {error, issuer_not_found} ->
            trusted_cert_and_path(CRL, issuer_not_found, {Db, DbRef}) 
     end;
-trusted_cert_and_path(CRL, issuer_not_found, {Db, DbRef} = DbInfo) -> 
+trusted_cert_and_path(CRL, issuer_not_found, {Db, DbInfo})-> 
     case find_issuer(CRL, DbInfo) of
        {ok, OtpCert} ->
-            {ok, Root, Chain} = ssl_certificate:certificate_chain(OtpCert, Db, DbRef),
+            {ok, Root, Chain} = ssl_certificate:certificate_chain(OtpCert, Db, DbInfo),
            {ok, Root, lists:reverse(Chain)};
          {error, issuer_not_found} ->
              {error, unknown_ca}

```

**Comment by IngelaAndin** [comment]
Fixed this issue and another refactor bug in PR-4601. I extended the test suite but I do not get exactly the same situation as in your manual test, that passes. If you have any tips for generating test certs and CRLs that are exactly equivalent we are interested. Otherwise, we will look into that later when we get more time to invest in this area.

**Comment by voltone** [comment]
I checked the CRL and certificate generation for differences with my own local test setup or the ericsson.com scenario, but nothing stands out. I haven't yet run the tests and looked at the files that were generated, though. So are you saying the test with DER certs in `cacerts` passes also without the patch, so it doesn't exercise the code path of this bug? If that's the case I'll look into it some more.

**Comment by IngelaAndin** [comment]
I force pushed the PR, it would be nice if you can verify it for your test cases aswell.

**Comment by voltone** [comment]
I'll try to have a look later, but I can't promise that I can get to it today...

**Comment by voltone** [comment]
Hmm, I'm getting new `function_clause` errors when running my local test suite, for different scenarios. I will try to isolate the error, but I gotta go to two long meetings now so it will be a few hours before I can continue. In the meantime I think it might be best not to merge just yet...

BTW, supporting AKI with KeyId, rather then issuer RDN & S/N, would be great for chain building in general. I think almost all public CAs use only KeyIds in their AKI extension these days.

**Comment by voltone** [comment]
Ok, so with the current PR I see `function_clause` errors for `ssl_crl:find_issuer/2`. This seems to fix it:

```diff
diff --git a/lib/ssl/src/ssl_crl.erl b/lib/ssl/src/ssl_crl.erl
index a527aba9c6..fcc9c4cfbc 100644
--- a/lib/ssl/src/ssl_crl.erl
+++ b/lib/ssl/src/ssl_crl.erl
@@ -59,7 +59,7 @@ search_certpath(CRL, CertPath, Db, DbRef) ->
     end.
 
 search_db(CRL, Db, DbRef)->
-    case find_issuer(CRL, DbRef) of
+    case find_issuer(CRL, {Db, DbRef}) of
        {ok, OtpCert} ->
             {ok, Root, Chain} = ssl_certificate:certificate_chain(OtpCert, Db, DbRef),
            {ok, Root, lists:reverse(Chain)};
```

**Comment by IngelaAndin** [comment]
Humm ..  thank you for testing. This patch does not work for the manual test. I think I partly I understand the confusion. I will make a new suggestion tomorrow and see if we can find a fully working solution.

**Comment by IngelaAndin** [comment]
I have made a new try to get this working again.  It works for me could you please try it too.
The new version of the branch should have sha: 6947aad9b8efb6f6a27a4b96fcb0bc84d9591ed7

**Comment by voltone** [comment]
My test suite passes too 🎉

**Comment by IngelaAndin** [comment]
Solved by #4601
